### PR TITLE
alignment-aware memory allocation

### PR DESCRIPTION
**Pull request by chenbd**
issues #1597

### CODE REVIEW

File: include/h2o/memory.h [hunk 1]
```diff
@@ -199,7 +199,17 @@ void h2o_mem_clear_pool(h2o_mem_pool_t *pool);
 /**
  * allocates given size of memory from the memory pool, or dies if impossible
  */
-void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz);
+void *h2o_mem_alloc_pool_aligned(h2o_mem_pool_t *pool, size_t alignment, size_t size);
+
+#ifdef H2O_NO_BUILT_IN_ALIGNOF
```
Comment (kazuho) [inline include/h2o/memory.h]: Please test the name of the compiler using `#if` here to determine if we can use `__alignof__`.

The header file will be used from other applications and that might not involve the invocation of CMake.

File: include/h2o/memory.h [hunk 2]
```diff
@@ -199,7 +199,17 @@ void h2o_mem_clear_pool(h2o_mem_pool_t *pool);
...
+#define H2O_ALIGNOF(type) (16)
+#else
+#define H2O_ALIGNOF(type) (__alignof__(type))
+#endif
+#define H2O_ALIGN(x, a) (((x) + (a)-1) & ~((a)-1))
```
Comment (kazuho) [inline include/h2o/memory.h]: I would appreciate it if we could refrain from adding a public interface to calculate alignment. My understanding is that it is only used in memory.c.

File: lib/common/memory.c [hunk 3]
```diff
@@ -139,10 +139,12 @@ void h2o_mem_clear_pool(h2o_mem_pool_t *pool)
     pool->chunk_offset = sizeof(pool->chunks->bytes);
 }
 
-void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz)
+void *h2o_mem_alloc_pool_aligned(h2o_mem_pool_t *pool, size_t alignment, size_t sz)
 {
     void *ret;
 
+    assert(alignment <= 2 * sizeof(void *) && sz % alignment == 0);
```
Comment (kazuho) [inline lib/common/memory.c]: Can we remove this assert?

I do not think that `alignment <= 2 * sizeof(void *)` is something that is guaranteed on every platform. `sz % alignment == 0` is going to be slow, and I do not think we need to test this considering how the function is used; it's essentially trying to detect a compiler bug in using the compiler itself.

Related to this, I think that we should eliminate `st_h2o_mem_pool_chunk_t::_dummy`. Instead, I would suggest changing the definition of `st_h2o_mem_pool_chunk_t` to something like below and start `h2o_mem_pool_t::chunk_offset` from `offsetof(union un_h2o_mem_pool_chunk_t, next)`.
```
union un_h2o_mem_pool_chunk_t {
    struct st_h2o_mem_pool_chunk_t *next;
    char bytes[4096];
};
```

File: lib/common/memory.c [hunk 4]
```diff
@@ -152,11 +154,9 @@ void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz)
     }
 
     /* return a valid pointer even for 0 sized allocs */
-    if (sz == 0)
-        sz = 1;
+    if (H2O_UNLIKELY(sz == 0))
+        sz = alignment;
```
Comment (kazuho) [inline lib/common/memory.c]: I think that `h2o_mem_alloc_pool(int, 0)` should return a unique pointer that is aligned for `int`. Hence keep the code as-is.

File: lib/common/memory.c [hunk 5]
```diff
@@ -347,7 +348,8 @@ void h2o_vector__expand(h2o_mem_pool_t *pool, h2o_vector_t *vector, size_t eleme
     while (vector->capacity < new_capacity)
         vector->capacity *= 2;
     if (pool != NULL) {
-        new_entries = h2o_mem_alloc_pool(pool, element_size * vector->capacity);
+        new_entries =
+            h2o_mem_alloc_pool_aligned(pool, 2 * sizeof(void *), H2O_ALIGN(element_size * vector->capacity, 2 * sizeof(void *)));
```
Comment (kazuho) [inline lib/common/memory.c]: How about passing the alignment requirement as an argument to `h2o_vector__expand`? Call to the function is wrapped via `h2o_vector_reserve` in which you can get the name of the typed element.

File: include/h2o/memory.h (hunk 1 above)
Comment (chenbd) [reply include/h2o/memory.h]: Do not `determining it in CMakeFiles.txt` sounds good.

i searched on the web and seems most compilers support `__alignof__`.
if we use `#if` to judge, i'm confused about the judge condition.
is there any platform  we need support but missing `__alignof__`?

File: include/h2o/memory.h (hunk 2 above)
Comment (chenbd) [reply include/h2o/memory.h]: yes, will be fixed later.

File: lib/common/memory.c (hunk 4 above)
Comment (chenbd) [reply lib/common/memory.c]: you mean  leave `sz = 1` as-is? 
seems we can do this.

even with `sz = alignment`,  `h2o_mem_alloc_pool(int, 0)` also return a unique pointer that is aligned for int.

File: lib/common/memory.c (hunk 5 above)
Comment (chenbd) [reply lib/common/memory.c]: fixed. please check commit 4dd533.

File: include/h2o/memory.h (hunk 1 above)
Comment (kazuho) [reply include/h2o/memory.h]: That's a very good question.

I now think that we do not need a check. Let's expect that `__alignof__` is always available. We already use things like `__builtin_clzll`; it is likely that the range of compilers we support will remain the same.

File: lib/common/memory.c (hunk 4 above)
Comment (kazuho) [reply lib/common/memory.c]: Thank you for the changes.

FTR the issue was that there was a case that left the `chunk_offset` unchanged leading to the uniqueness not being guaranteed; for example when `chunk_offset` is a multiple of 4 and `h2o_mem_alloc_pool(pool, uint32_t, 0)` is called.

File: lib/common/memory.c [hunk 6]
```diff
@@ -106,7 +107,7 @@ void h2o_mem_free_recycle(h2o_mem_recycle_t *allocator, void *p)
 void h2o_mem_init_pool(h2o_mem_pool_t *pool)
 {
     pool->chunks = NULL;
-    pool->chunk_offset = sizeof(pool->chunks->bytes);
+    pool->chunk_offset = sizeof(*pool->chunks);
```
Comment (kazuho) [inline lib/common/memory.c]: Would you mind elaborating why the change is necessary? I see that you have changed all the `sizeof`s related to the calculation of the buffer size.

My understanding is that `pool->chunk_offset` has been (and can continue to be) a offset for `(st|un)_h2o_mem_pool_chunk_t::bytes`, and I do not understand why you need to change.

File: lib/common/memory.c [hunk 7]
```diff
@@ -152,27 +153,26 @@ void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz)
...
         /* allocate new chunk */
-        struct st_h2o_mem_pool_chunk_t *newp = h2o_mem_alloc_recycle(&mempool_allocator, sizeof(*newp));
+        union un_h2o_mem_pool_chunk_t *newp = h2o_mem_alloc_recycle(&mempool_allocator, sizeof(*newp));
         newp->next = pool->chunks;
         pool->chunks = newp;
-        pool->chunk_offset = 0;
+        pool->chunk_offset = sizeof(union un_h2o_mem_pool_chunk_t *);
     }
 
+    pool->chunk_offset = H2O_ALIGN(pool->chunk_offset, alignment);
```
Comment (kazuho) [inline lib/common/memory.c]: Do we need to make this calculation before `if (sizeof(*pool->chunks) - pool->chunk_offset < sz)`?

File: lib/common/memory.c [hunk 8]
```diff
@@ -132,18 +133,18 @@ void h2o_mem_clear_pool(h2o_mem_pool_t *pool)
...
+    pool->chunk_offset = sizeof(*pool->chunks);
 }
 
-void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz)
+void *h2o_mem_alloc_pool_aligned(h2o_mem_pool_t *pool, size_t alignment, size_t sz)
 {
     void *ret;
 
-    if (sz >= sizeof(pool->chunks->bytes) / 4) {
+    if (sz >= sizeof(*pool->chunks) / 4) {
```
Comment (kazuho) [inline lib/common/memory.c]: Maybe this should be `sz >= (sizeof(pool->chunks->bytes) - sizeof(pool->chunks->next)) / 4`?

File: lib/common/memory.c [hunk 9]
```diff
@@ -152,27 +153,26 @@ void *h2o_mem_alloc_pool(h2o_mem_pool_t *pool, size_t sz)
...
-    sz = (sz + 15) & ~15;
-    if (sizeof(pool->chunks->bytes) - pool->chunk_offset < sz) {
+    if (sizeof(*pool->chunks) - pool->chunk_offset < sz) {
         /* allocate new chunk */
-        struct st_h2o_mem_pool_chunk_t *newp = h2o_mem_alloc_recycle(&mempool_allocator, sizeof(*newp));
+        union un_h2o_mem_pool_chunk_t *newp = h2o_mem_alloc_recycle(&mempool_allocator, sizeof(*newp));
         newp->next = pool->chunks;
         pool->chunks = newp;
-        pool->chunk_offset = 0;
+        pool->chunk_offset = sizeof(union un_h2o_mem_pool_chunk_t *);
```
Comment (kazuho) [inline lib/common/memory.c]: Can we change the sizeof to `sizeof(newp->next)`?

Relying on the variable name is more prone to errors than copy-pasting the type of the variable name.

File: lib/common/memory.c (hunk 6 above)
Comment (chenbd) [reply lib/common/memory.c]: because now un_h2o_mem_pool_chunk_t is a union, and if offsetof(un_h2o_mem_pool_chunk_t, bytes) will be 0 which differs from struct.

File: lib/common/memory.c (hunk 8 above)
Comment (chenbd) [reply lib/common/memory.c]: yes.
we can change this if we want **exactly** the same as before.

File: lib/common/memory.c (hunk 9 above)
Comment (chenbd) [reply lib/common/memory.c]: yes, will be fixed.

File: lib/common/memory.c (hunk 7 above)
Comment (chenbd) [reply lib/common/memory.c]: yes, we need make `chunk_offset` align before judge, also when new chunk is allocated.

File: lib/common/memory.c (hunk 6 above)
Comment (kazuho) [reply lib/common/memory.c]: Thank you for the answer.

I have switched the code back to use `sizeof ...bytes`. In addition to being more natural (as stated in my comment above) it does not require the knowledge that 1) `bytes` is the largest member of a 2) union.

**Comment by kazuho** [comment]
Thank you for your work on the issue! I have merged the pull request after making some cosmetic changes. I think that all the changes are fine, but please let me know if you have any concern.
